cli/lib/tasks/process-js-task.js

```diff
@@ -0,0 +1,381 @@
+		if (fullBuild) {
+			return this.doFullTaskRun();
+		}
+
+		this.jsBootstrapFiles.splice(0, 0, ...this.data.jsBootstrapFiles);
+
+		const deletedFiles = this.filterFilesByStatus(changedFiles, 'deleted');
+		const deletedPromise = Promise.all(deletedFiles.map(filePath => limit(() => this.handleDeletedFile(filePath))));
+
+		const updatedFiles = this.filterFilesByStatus(changedFiles, [ 'created', 'changed' ]);
+		const updatedPromise = Promise.all(updatedFiles.map(filePath => limit(() => this.processJsFile(filePath))));
+
+		// @fixme: can be removed in 9.0 to even further decrease build times on incremental builds
+		const unchangedFiles = Array.from(this.inputFiles).filter(filePath => !changedFiles.has(filePath));
+		const unchangedPromise = Promise.all(unchangedFiles.map(filePath => limit(() => this.processJsFile(filePath))));
+
+		return Promise.all([ deletedPromise, updatedPromise, unchangedPromise ]);
+	}
+
+	/**
+	 * Loads last run's task data if this task will be skipped entirely.
+	 *
+	 * Note that this function will be called when no input files changed. Since
+	 * we also depend on some config values from the builder this is used to
+	 * fallback to a full task run if required.
+	 *
+	 * @return {Promise}
+	 */
+	async loadResultAndSkip() {
+		const loaded = await this.loadTaskData();
+		const fullBuild = !loaded || this.requiresFullBuild();
+		if (fullBuild) {
+			return this.doFullTaskRun();
+		}
+
+		Object.keys(this.data.jsFiles).forEach(relPath => this.builder.unmarkBuildDirFile(this.data.jsFiles[relPath].dest));
+	}
+
+	/**
+	 * Function that will be called after the task run finished.
+	 *
+	 * Used to update the task data before it gets written to disk.
+	 */
+	async afterTaskAction() {
+		await super.afterTaskAction();
+
+		this.data.jsBootstrapFiles = this.jsBootstrapFiles;
+		this.data.jsFiles = this.jsFiles;
+		this.data.analyzeOptionsHash = this.generateHash(JSON.stringify(this.defaultAnalyzeOptions));
+		await this.saveTaskData();
+
+		this.builder.unmarkBuildDirFiles(this.incrementalDirectory);
+	}
+
+	/**
+	 * Creates the hooks that are required during JS processing.
+	 */
+	createHooks() {
+		let compileJsFileHook = this.builder.cli.createHook('build.ios.compileJsFile', this.builder, (r, from, to, done) => {
+			// Read the possibly modified file contents
+			const source = r.contents;
+
+			// If the file didn't change from previous run, return early
+			const currentHash = this.generateHash(source);
+			const previousHash = this.data.contentHashes[from];
+			if (previousHash && previousHash === currentHash) {
+				this.builder.unmarkBuildDirFile(to);
+				return done();
+			}
+
+			this.transformAndCopy(source, from, to).then(() => {
+				this.data.contentHashes[from] = currentHash;
+				return done();
+			}).catch(e => {
+				// if we have a nicely formatted pointer to syntax error from babel, print it!
+				if (e.codeFrame) {
+					this.logger.error(e.codeFrame);
+				}
+
+				done(e);
```

Comment:
<!--
  0 failure: 
  1 warning:  cli/lib/tasks/pro...
  
  
  DangerID: danger-id-Danger;
  File: cli/lib/tasks/process-js-task.js;
  Line: 156;
-->

- :warning: cli/lib/tasks/process-js-task.js line 156 – Avoid calling back inside of a promise. (promise/no-callback-in-promise)